pyop2/compilation.py

```diff
@@ -355,99 +383,127 @@ def get_so(self, jitmodule, extension):
-
-    :arg cppargs: A list of arguments to pass to the C compiler
-         (optional).
-    :arg ldargs: A list of arguments to pass to the linker (optional).
-
-    :arg cpp: Are we actually using the C++ compiler?
+class MacClangCompiler(Compiler):
+    """A compiler for building a shared library on mac systems."""
+    _name = "Mac Clang"
 
-    :kwarg comm: Optional communicator to compile the code on (only
-        rank 0 compiles code) (defaults to COMM_WORLD).
-    """
+    _cflags = ["-fPIC", "-Wall", "-framework", "Accelerate", "-std=gnu11"]
+    _cxxflags = ["-fPIC", "-Wall", "-framework", "Accelerate"]
+    _ldflags = ["-dynamiclib"]
 
-    def __init__(self, cppargs=[], ldargs=[], cpp=False, comm=None):
+    @property
+    def _optflags(self):
         machine = platform.uname().machine
         opt_flags = ["-O3", "-ffast-math"]
         if machine == "arm64":
             # See https://stackoverflow.com/q/65966969
             opt_flags.append("-mcpu=apple-a14")
         elif machine == "x86_64":
             opt_flags.append("-march=native")
+        return opt_flags
 
-        if configuration["debug"]:
-            opt_flags = ["-O0", "-g"]
-
-        cc = "mpicc"
-        stdargs = ["-std=gnu11"]
-        if cpp:
-            cc = "mpicxx"
-            stdargs = []
-        cppargs = stdargs + ['-fPIC', '-Wall', '-framework', 'Accelerate'] + \
-            opt_flags + cppargs
-        ldargs = ['-dynamiclib'] + ldargs
-        super(MacCompiler, self).__init__(cc,
-                                          cppargs=cppargs,
-                                          ldargs=ldargs,
-                                          cpp=cpp,
-                                          comm=comm)
-
-
-class LinuxCompiler(Compiler):
-    """A compiler for building a shared library on linux systems.
-
-    :arg cppargs: A list of arguments to pass to the C compiler
-         (optional).
-    :arg ldargs: A list of arguments to pass to the linker (optional).
-    :arg cpp: Are we actually using the C++ compiler?
-    :kwarg comm: Optional communicator to compile the code on (only
-    rank 0 compiles code) (defaults to COMM_WORLD)."""
-    def __init__(self, cppargs=[], ldargs=[], cpp=False, comm=None):
-        opt_flags = ['-march=native', '-O3', '-ffast-math']
-        if configuration['debug']:
-            opt_flags = ['-O0', '-g']
-        cc = "mpicc"
-        stdargs = ["-std=gnu11"]
-        if cpp:
-            cc = "mpicxx"
-            stdargs = []
-        cppargs = stdargs + ['-fPIC', '-Wall'] + opt_flags + cppargs
-        ldargs = ['-shared'] + ldargs
-
-        super(LinuxCompiler, self).__init__(cc, cppargs=cppargs, ldargs=ldargs,
-                                            cpp=cpp, comm=comm)
+    _debugflags = ["-O0", "-g"]
+
+
+class LinuxGnuCompiler(Compiler):
+    """The GNU compiler for building a shared library on Linux systems."""
+    _name = "GNU"
+
+    _cflags = ["-fPIC", "-Wall", "-std=gnu11"]
```

Comment:
I'm not a big fan of these being private attributes but it's probably the best solution.

Reply:
I'm happy to take on an alternative suggestion, this is mainly so the cflags getter method works correctly. One alternative is hard coding them into the methods, but I'm even less of a fan of that!